### DISCUSSION TITLE
Usage of useShallow with nested objects

### DISCUSSION THREAD

**Discussion by brunano21** [pmndrs/zustand#2316] [Q&A]
Hi there,

in one of my React apps, I have decided to give Zustand a shot. The migration from Context APIs was quite simple even if the app is quite complex in terms of state management.

The reasons behind this question are how to properly use `useShallow` with nested objects and to understand if I'm on the right track or if I'm just overthinking.

Here I want to share what the initial store looks like:

```app-store.js
const initialState = {
    items: [],
    ...
    error: null,
    user: null,
};
```
The `app-store` holds the logged user's details, and other less-important info. What's relevant here is the `items` array: it contains a set of objects with mixed types: primitives and complex objects (nested arrays most of the time).
Each item is like the one below - for simplicity, I'm displaying only 2 properties along with the id and the status.
```js
{ 
   id, // string
   status, // string
   pages: [{ id, createdAt, ...}, { ... }], 
   comms: [{ id, createdAt, content, ...}, { ... }], 
   ...
}
```

Let's assume that each property of the item (comms, pages) is used to render a small portion of the same page - let's call them _PagesSection_ and _CommsSection_ (they are React functional components). 

I intend to avoid re-rendering the PagesPortion when no changes have occurred to that `pages` array (pages can only grow in length, once a `page` is pushed in its array, its data won't ever change. The same applies to the comms (same assumptions).

The current _PagesSection_ component looks like:

```js
function PagesSection() {
    const { id } = useParams();

    const { pages, status, item } = useAppStore(
        (state) => {
            const item = state.items.find(({id: itemId}) => itemId === id);
            return {
                pages: item.pages,
                status: item.status,
                item,
            };
        },
        (prev, curr) => {
            const condition =
                prev.pages?.length === curr.pages?.length &&
                prev.status === curr.status &&
                prev?.item.id === curr.item.id;

            return condition;
        },
    );

   ...
   // rest of the component where pages and status will be used to render the component itself. 
   ...
}
```

I'm aware that the above is now marked as deprecated - a warning shows up in my IDE with the following: _@deprecated — Use createWithEqualityFn from 'zustand/traditional'_. So, [looking for alternatives](https://docs.pmnd.rs/zustand/guides/prevent-rerenders-with-use-shallow), I'm thinking about whether `useShallow` can help in this case - the intention as anticipated above is to avoid to re-render this component in case there is no change to the item's status or the item's pages.

Please note that in components similar to PagesSection, the equality function makes usage of `deep-Equal. See example below:

```js
const { incident, roles, status } = useAppStore(
     (state) => {
         const item = state.items.find(({id: itemId}) => itemId === id);
            return {
                item,
                roles: item.roles,
                status: item.status,
            };
        },
     (prev, curr) => {
           const condition =
                deepEqual(prev.roles, curr.roles) &&
                prev.status === curr.status &&
                prev?.item.id === curr.item.id;

            return condition;
        },
);
```

Any advice and feedback is highly appreciated!

**Comment by dai-shi** [comment]
If you are using custom equality function other than `shallow`.
Using `zustand/traditional` is fine and it will stay in Zustand v5.
If your case works with `shallow`, then `useShallow` works too.

> the equality function makes usage of `deep-Equal.

So, in that case, `useShallow` doesn't work. However, as I understand from the code, you could update `roles` only if changed, and then you could use `useShallow`.
```js
function updateRoles(...) {
  const oldRoles = item.roles;
  const newRoles = oldRoles.filter(...); // or something
  if (deepEqual(oldRoles, newRoles)) {
    return oldRoles;
  } else {
    return newRoles;
  }
}
```

**Reply by brunano21** [reply]
Thanks for the quick reply, sir!

The current `updateRoles` is something like this - (I'm using `immer` also): 
```
updateRoles: (id, roles, ...) => {
        const idx = get().items.findIndex((i) => i.id === id);
        set((state) => {
            state.items[idx].roles = roles;
        });
    },
```

The `updateRoles` function is called as a side-effect of an API call and as such old and new `roles` content are always different from each other.

What I'd like to understand is how to replace the deprecated `useStore(...., equalityFn)` when creating a slice, aimed to reduce the number of re-renders if the `roles` array doesn't change.

Also, in regards to pages/comms arrays - if their length increases and I use 'useShallow' it can happen that the associated component won't re-render because the previous and current piece of states are shallowly the same - right? if so, how can pass an equality function to make sure this doesn't happen?

**Reply by dai-shi** [reply]
> the deprecated useStore(...., equalityFn)

It's deprecated, but it's in the new `zustand/traditonal` entry point.
So, I wonder if your problem is migrating `zustand/traditional`, or you want to avoid `zustand/traditional`.

**Comment by dai-shi** [comment]
> The current `updateRoles` is something like this - (I'm using `immer` also):
> 
> ```
> updateRoles: (id, roles, ...) => {
>         const idx = get().items.findIndex((i) => i.id === id);
>         set((state) => {
>             state.items[idx].roles = roles;
>         });
>     },
> ```

That doesn't seem correct unless you are using `immer` middleware.

Here's something to avoid deepEqual in hooks.
```js
    updateRoles: (id, roles, ...) => {
        const idx = get().items.findIndex((i) => i.id === id);
        set((state) => {
            if (deepEqual(state.items[idx].roles, roles)) {
                return state;
            }
            const prevItem = state.items[idx]
            const newItem = { ...prevItem, roles };
            return {
                items: [...state.items.slice(0, idx), newItem, ...state.items.slice(idx + 1)],
            };
        });
    },
```

**Reply by brunano21** [reply]
> That doesn't seem correct unless you are using immer middleware.

Yes, I'm using immer.

**Reply by dbritto-dev** [reply]
```ts
    updateRoles: (id, roles, ...) => {
        set((state) => {
            const idx = state.items.findIndex((i) => i.id === id);
            if (deepEqual(state.items[idx].roles, roles)) {
                return state;
            }
            const prevItem = state.items[idx]
            const newItem = { ...prevItem, roles };
            return {
                items: [...state.items.slice(0, idx), newItem, ...state.items.slice(idx + 1)],
            };
        });
    },
```
